### ISSUE TITLE
@extend behaving illogically

### ISSUE THREAD

**Issue by Tokimon** [sass/sass#1428]
**Version**: 3.4.3

**Description of issue**
When extending an %extend-only extension with a multilevel selector, that should be descendant of an arbitrary class, the selector is extended in an illogical way, where the parent class is placed at a random place inside the multilevel selector.

**Example code**

``` sass
@mixin test() {
    .test3 &  {
        background: #ff0;
    }
}

.test {
    %extend {
        background: #f0f;
    }
}

%extend2 {
    .test2 &  {
        background: #f00;
    }
}

%extend3 {
    @include test();
}

a[href] span b em {
    .test4 & {
        background: #00f;
    }

    @extend %extend;
    @extend %extend2;
    @extend %extend3;
}
```

**Produced result**

``` css
.test a[href] span b em, a[href] span b .test em {
  background: #f0f;
}

.test2 a[href] span b em, a[href] span b .test2 em {
  background: #f00;
}

.test3 a[href] span b em, a[href] span b .test3 em {
  background: #ff0;
}

.test4 a[href] span b em {
  background: #00f;
}
```

**Expected result**

``` css
.test a[href] span b em {
  background: #f0f;
}

.test2 a[href] span b em {
  background: #f00;
}

.test3 a[href] span b em {
  background: #ff0;
}

.test4 a[href] span b em {
  background: #00f;
}
```

**Explanation**
I created 4 ways to do the same thing and only the `.test4` part gives the result I was looking for, when extending. As you can see it places the class `.testX` inside the multilevel selector after the `b`, whereas the expected behavior is to replace `&` in `.testX &` with the selector resulting in something like `.test2 a[href] span b em`. This is a strange behavior (probably caused by the fact it treats the selector as a list) that in the end generates extra unwanted selectors and basically makes the use of extend pretty limited.

**Final note**
I have tried experimenting with the number of levels of the selector (to something like this  `p q a[href] span b i em`) but it seems it always inserts the parent class before the last part of the selector (`em`). I am unsure if this is the desired effect or not, but to me it seems illogical. it is like it can't decide what `&` really refers to. Again to me the logical result would be the same as when you nest it, like I did with `.test4`.

**Comment by lolmaus** [comment]
I believe this is intentional.

You're doing pretty complicated extends there, and Sass is trying to produce all possible combinations because there is no way for it to know which combination will appear in HTML.

**Comment by Tokimon** [comment]
I don't understand how it is complicated. Yes I have a lot of extends in my selector, but other that is just to have several examples of ways to do it.

It is a pretty basic extend, just the selector has more than one level.
In theory it should be the same as doing (in my `.test2` example):

``` sass
a[href] span b em {
    .test2 &  {
        background: #f00;
    }
}
```

And that exact phrase work fine (as you can see with the `.test4` example), but if I do it as an extend, it doesn't work. That is what puzzles me.

**Comment by lolmaus** [comment]
What exactly does not work? Please provide one specific example.

**Comment by Tokimon** [comment]
Ok.
This:

``` sass
a[href] span b em {
    .test4 & {
        background: #00f;
    }
}
```

Gives the result I am looking for:

``` css
.test4 a[href] span b em {
        background: #00f;
}
```

But if I extend it like this:

``` sass
%extend {
    .test4 & {
        background: #00f;
    }
}

a[href] span b em {
    @extend %extend;
}
```

It gives me this double result like this:

``` css
.test4 a[href] span b em, a[href] span b .test4 em {
        background: #00f;
}
```

I just don't understand why it would choose to add the last selector (`a[href] span b .test4 em`), as it doesn't do that in nesting and my understanding of the `%extend-only` feature, is that it acts as a sort of placeholder, where all your selectors is put. A bit like:

``` sass
%extend {
    .test4 & {
        background: #00f;
    }
}

.module1 { @extend %extend; }
.module2 h2 { @extend %extend; }
.module3 a[href] span b em { @extend %extend; }
```

Would be the same as doing:

``` sass
.module1,
.module2 h2,
.module3 a[href] span b em {
    .test4 & {
        background: #00f;
    }
}
```

But since i often spread my components into different files, it is much more convenient to be able to do the extend, instead of maintaining a separate list in a 3rd file. Especially if I am including different modules in different files like:

_page1.css_ includes module1 + module2
and therefore the extended list should be:

``` sass
.module1,
.module2 h2 {
    .test4 & {
        background: #00f;
    }
}
```

_page2.css_ includes module2 + module3
and therefore the extended list should be:

``` sass
.module2 h2,
.module3 a[href] span b em {
    .test4 & {
        background: #00f;
    }
}
```

_unique-page.css_ includes module1 + module3
and therefore the extended list should be:

``` sass
.module1,
.module3 a[href] span b em {
    .test4 & {
        background: #00f;
    }
}
```

This should be doable with the use of the `%extend-only` feature, but right now this is, in my opinion, a bit unusable for that, as it creates the double selector.

**Comment by nex3** [comment]
**tl;dr: This is intended.**

Let's look at a simpler case to understand why this works the way it does:

``` scss
.a %b {color: blue}
.c .d {@extend %b}
```

First of all, what does that `@extend` mean? It means "style all elements that match `.d` and have a parent that matches `.c` as though they also matched `%b`." The only thing we know about how elements matching `%b` are styled is that, if they have a parent that matches `.a`, they have blue text.

Okay, great. How do we combine these? Well, if an element matches `.d`, has a parent that matches `.c`, and has a parent that matches `.a`, it should have blue text. But we don't know anything about how those two parents—the one that matches `.c` and the one that matches `.a`—should be ordered. Both of these spans should have blue text:

``` html
<div class="a"><div class="c"><span class="d">Text</span></div></div>
<div class="c"><div class="a"><span class="d">Text</span></div></div>
```

Thus, the CSS we have to generate is the following:

``` css
.a .c .d, .c .a .d {
  color: blue;
}
```

This is where those "extra" selectors are coming from.

**Comment by Tokimon** [comment]
If that is what is intended, then that is that.
But I still don't agree with the behavior because when you extend a selector you extend the whole selector not just the last part of it. Take your example for instance. It should still replace `%b` with the whole selector and not consider whether or not I meant the last part or the whole part. So:

``` sass
.a %b {color: blue}
.c .d {@extend %b}
```

should still just replace `%b` for a selector like this:

``` sass
.a .c .d {color: blue}
```

In this case I would not be extending `.d`but the whole selector `.c .d`, so it should not be in doubt of how the HTML is. Just insert it in the place of the placeholder/extend-only selector. This is also true if we extended a class or other selectors:

``` sass
.a .b {color: blue}
.c .d {@extend .b}
```

This is should be the same as:

``` sass
.a .b,
.a .c .d {color: blue}
```

Even if you do a nesting, it should still extend the whole selector, not the selector and another arbitrary "in-case-you-actually-ment-this" selector. No if I wanted to extend `.d` as a child of `.a` without `.c`, then I would do this:

``` sass
.a %b {color: blue}
.d {@extend %b}
.c .d {@extend %b}
```

And if I wanted the `.c .a .d` selector I would do it manually like this:

``` sass
%e { color: blue }
.a %b {color: blue}

.c {
    .d { @extend %b  }
    .a .d { @extend %e }
}
```

Having SASS guess what I mean can lead to unwanted selectors and/or unintended behaviors. Most of the time you don't need the extra selector and it is just annoying. I have an actual case:

``` sass
.theme-somename %theme-background { background: $theme-color; }
abbr[title] span { @extend %theme-background; }
```

This resulted in a selector like this:

``` sass
.theme-somename abbr[title] span,
abbr .theme-somename[title] span { background: $theme-color; }
```

Which is definitely not what I wanted. I just wanted the first selector. The second is never used and therefore just takes up space in an already large stylesheet.

Bottom line is. I love SASS and I have been a happy user of it for quite some time now. I know very well most aspects of it and love it for it. But this behavior strikes me as a bit odd and, to me, illogical. especially because it takes the intended selector and makes another from the last item in the selector, which might not be where I would want it. So I unerstand it is a help, but more often thatn not it is annoying.
You might not agree with me, but I know I am not the only one who find it a strange behavior.

**Comment by nex3** [comment]
You're still thinking of `@extend` in terms of how it manipulates text rather than how it affects styling. An argument from the perspective of text manipulation isn't going to be persuasive, because that's not the plane on which `@extend` operates. If you can make a semantic argument instead, you might make more progress.
